miio/gateway/gateway.py

```diff
@@ -387,3 +407,86 @@ def get_illumination(self):
             raise GatewayException(
                 "Got an exception while getting gateway illumination"
             ) from ex
+
+    def Register_callback(self, id, callback):
+        """Register a external callback function for updates of this subdevice."""
+        if id in self._registered_callbacks:
+            _LOGGER.error(
+                "A callback with id '%s' was already registed, overwriting previous callback",
+                id,
+            )
+        self._registered_callbacks[id] = callback
+
+    def Remove_callback(self, id):
+        """Remove a external callback using its id."""
+        self._registered_callbacks.pop(id)
+
+    def gateway_push_callback(self, action, params):
+        """Callback from the push server regarding the gateway itself."""
+        for callback in self._registered_callbacks.values():
+            callback(action, params)
+
+    def push_callback(self, source_device, action, params):
+        """Callback from the push server."""
+        if source_device not in self.devices:
+            if source_device == f"lumi.{self.did}":
+                self.gateway_push_callback(action, params)
+                return
+
+            _LOGGER.error(
+                "'%s' callback from device '%s' not from a known device",
+                action,
+                source_device,
+            )
+            return
+
+        device = self.devices[source_device]
+        device.push_callback(action, params)
+
+    def install_script(self, script_id, script_data):
+        """Install script such that the gateway will start pushing data for that
+        script."""
+        if self._push_server is None:
+            _LOGGER.error("Can not install script withouth a push_server")
+            return
+
+        if script_id in self._script_ids:
+            _LOGGER.error(
+                "Script_id '%s' already installed, not installing another time",
+                script_id,
+            )
+            return
+
+        self._script_ids.append(script_id)
+
+        return self.send(
+            "send_data_frame",
+            {
+                "cur": 0,
+                "data": script_data,
+                "data_tkn": 29576,
+                "total": 1,
+                "type": "scene",
+            },
+        )
+
+    def delete_script(self, script_id):
+        """Delete script by id."""
+        result = self.send("miIO.xdel", [script_id])
```

Comment:
both `gateway.send("miIO.xgetKeys")` and `gateway.send("miIO.xget")` return:
`miio.exceptions.DeviceError: {'code': -32601, 'message': 'Method not found.'}`

Reply:
I have been wondering the same thing, but have not found a way yet to get the installed "scripts" from a device

Reply:
Okay, I just saw those commands somewhere, so I was wondering if they are common across devices. But it may very well be, that that information is just kept on the server-side.